modules/javafx.graphics/src/main/native-prism-d3d/hlsl/vs2ps.h

```diff
@@ -23,30 +23,28 @@
  * questions.
  */
 
-struct ObjectPsIn {
-    float2  texD    : texcoord0;
-    float4  ambient : texcoord1;
-};
-
-struct LocalBump {
+struct PsInput {
 
     static const float nLights = 3;
 
-    float3 eye                : texcoord2;
-    float4 lights[nLights]    : texcoord3; // 3, 4, 5 [6]
-    float4 lightDirs[nLights] : texcoord7; // 7, 8, 9 [10]
-//  float3 debug              : texcoord11;
-};
+    // projection space = homogeneous clip space
+    float4 projPos : position; // must be outputed even if unused
 
-struct LocalBumpOut {
-    float4  pos  : position;
-    float   oFog : fog;
+//  needed for pixel lighting
+//  float3 worldPos           : texcoord1;
 
-    LocalBump lBump;
-};
+    float3 worldVecToEye               : texcoord2;
+    float3 worldVecsToLights[nLights]  : texcoord3; // 3, 4, 5
+    float3 worldNormLightDirs[nLights] : texcoord6; // 6, 7, 8
```

Comment:
This is an existing behavior, and I just want to open it for discussion and bring it to your consideration for future changes.
These are per vertex attributes and they may cause un-required processing/overhead during rasterization.
In a regular scenario where only one light i.e. default light is set, the other two values in these arrays are reset and are unused in pixel shader. But rasterizer interpolates these values for all pixel pass as input to pixel shader.
May be in future we can find a way to avoid this.

Reply:
I think that when the strict limit of 3 lights is removed then we will touch this as well.